### ISSUE TITLE
Unreachable host generates misleading exception when mutation is executed

### ISSUE THREAD

**Issue by fabis94** [zino-app/graphql-flutter#600]
**Describe the bug**
When invoking the `runMutation` callback of a Mutation, if the URI used in the `HttpLink` is unreachable due to an unreachable host, the exception that actually bubbles out of `runMutation` is this:

```
E/flutter (22479): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: NoSuchMethodError: The getter 'host' was called on null.
E/flutter (22479): Receiver: null
E/flutter (22479): Tried calling: host
E/flutter (22479): #0      Object.noSuchMethod  (dart:core-patch/object_patch.dart:53:5)
E/flutter (22479): #1      translateNetworkFailure 
package:graphql/…/exceptions/io_network_exception.dart:14
E/flutter (22479): #2      translateFailure 
package:graphql/…/exceptions/exceptions.dart:13
E/flutter (22479): #3      new HttpLink.<anonymous closure>.onListen 
```

The real underlying exception that gets thrown is a `SocketException` w/ a null `address` and the following message: "Failed host lookup: 'invalid-hostname.test'"

The misleading exception gets thrown, because the `translateNetworkFailure` method of this lib, tries to use the null `address` to get the hostname.

![image](https://user-images.githubusercontent.com/938316/79074076-07530580-7cf3-11ea-82c5-5247aa2aba48.png)

**To Reproduce**
1. Make sure your HttpLink URI has an invalid/nonexistent hostname
2. Execute any mutation (probably query as well)

**Expected behavior**
The exception should accurately tell you that the GraphQL client couldn't connect to the specified UI because of the invalid hostname. And from what I understand it shouldn't even be thrown, but it should be set as part of the result object.

**Comment by gouravd** [comment]
Any updates on this?

**Comment by nicolailinhares** [comment]
The same error happen in queries as well, is not specific to mutations.
The translateNetworkFailure crashes when you have the following error number in the Socket Exception:
SocketException: Failed host lookup: 'stg.api.biiz.com.br' (OS Error: No address associated with hostname, errno = 7)

A simple suggestion would create a NetworkException with null uri when the address is nil. Other simple solution would be to create another class of exception for cases when there is no address.

**Comment by aleksandr-plujnik** [comment]
@fabis94, @gouravd i think it's fixed in the 3.1.0 version

**Comment by micimize** [comment]
This is likely solved at least by `4.0.0-beta`. If anyone still has this issue please comment and I'll reopen.

**Comment by AlexKenbo** [comment]
We connect over a web socket with a data subscription. And all the data is successfully received.

But when we turn off the Internet connection in the emulator and do not turn on the Internet for more than 5 seconds.

Then after turning on the Internet, this error appears.

And we can't catch her in any way, because we don't know where to do it.

[ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: WebSocketChannelException: WebSocketChannelException: SocketException: Failed host lookup: '{MY_GQL_ENPOINT}' (OS Error: No address associated with hostname, errno = 7)
